### PR TITLE
Enable message pump strategies to light up on any framework that has the correct types

### LINKED ISSUE #3552: MessagePumpStrategy does not work for WPF on netcoreapp3.0 and upwards
I am trying port our applications to netcoreapp3.1.
During this i also tried to run our unit tests for that target framework.
To my surprise the tests deadlocked as soon as they use async or things like `Assert.CatchAsync`.
The issue boils down to `MessagePumpStrategy` which had special cases for windows forms and WPF on net framework 4.

Is there a reason why there are no special cases for netcoreapp3.0 and upwards or was it just an oversight and those should be there?

### PR DESCRIPTION
**Pull request by jnm2**
Fixes #3552

### CODE REVIEW

File: src/NUnitFramework/framework/Internal/MessagePumpStrategy.cs [hunk 1]
```diff
@@ -65,18 +62,52 @@ public override void WaitForCompletion(AwaitAdapter awaiter)
...
+
+                    var applicationRun = (Action)applicationType
+                        .GetMethod("Run", BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly, null, Type.EmptyTypes, null)
+                        .CreateDelegate(typeof(Action));
+
+                    var applicationExit = (Action)applicationType
+                        .GetMethod("Exit", BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly, null, Type.EmptyTypes, null)
+                        .CreateDelegate(typeof(Action));
+
+                    _instance = new WindowsFormsMessagePumpStrategy(applicationRun, applicationExit);
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/MessagePumpStrategy.cs:94]: There is no need for racing threads to agree upon a single WindowsFormsMessagePumpStrategy instance, so Interlocked.CompareExchange wouldn't add value.

**Comment by batzen** [comment]
@jnm2 I can confirm that my tests don't deadlock anymore after using the version from this PR.
Thank you so much!

**Comment by mikkelbu** [comment]
@jnm2 I can take a look at this PR this weekend. Currently, it does not build on macOS and Linux with the failure
```
...
Windows is required to build Windows desktop applications.
...
```

**Comment by batzen** [comment]
Looking at the code the changes in the project file don't seem necessary as the problem/issue is solved by using reflection.

**Comment by jnm2** [comment]
@batzen The changes (or something like them) are needed in order to make CI fail if support for .NET Core was revoked.

**Comment by batzen** [comment]
What do you mean with"revoked"? If we need something to make the build fail if something in .net changes i could provide some tests through a separate PR.

**Comment by jnm2** [comment]
@batzen I mean, I wanted to have a test that would fail before making any implementation changes. That way if we accidentally revert the behavior in the implementation (say, adding incorrect preprocessor back in), we will notice what we are breaking.

**Comment by jnm2** [comment]
I think Microsoft.NET.Sdk.WindowsDesktop is a no-go. When I have the time, I'll try to see what the next best way is to reference System.Windows.Forms and WindowsBase in the test project for .NET Core 3.1.

If necessary we can continue with no tests, but then there's nothing to tell us if we won't accidentally break one of the things that these tests guarantee on just .NET Core.

File: src/NUnitFramework/tests/SynchronizationContextTests.cs [hunk 2]
```diff
@@ -104,7 +104,7 @@ public static void AsyncDelegatesAreExecutedOnStaThread(AsyncExecutionApiAdapter
             });
         }
 
-#if NETFRAMEWORK
+#if !NETCOREAPP2_1
```
Comment (jnm2) [inline src/NUnitFramework/tests/SynchronizationContextTests.cs:107]: This change is what is causing the CI build failure. `System.Windows.Threading.DispatcherSynchronizationContext` and `System.Windows.Threading.DispatcherSynchronizationContext` aren't available to compile or run against in .NET Core 3.1 unless we use the Windows desktop SDK (which fails the build on other OSs) or come up with a workaround.

File: src/NUnitFramework/tests/nunit.framework.tests.csproj [hunk 3]
```diff
@@ -38,6 +38,9 @@
 
   <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
     <Reference Include="System.Web" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' or '$(TargetFramework)' != 'netcoreapp3.1'">
     <Reference Include="System.Windows.Forms" />
     <Reference Include="WindowsBase" />
   </ItemGroup>
```
Comment (jnm2) [inline src/NUnitFramework/tests/nunit.framework.tests.csproj:46]: These are currently producing missing reference warnings during compilation.

**Comment by batzen** [comment]
I know how we can referecence the correct sdk only on windows. Will post code later or sometime during the weekend.

**Comment by jnm2** [comment]
@batzen That would be great! I reset the branch back to the version that has the two csprojs on the Windows desktop SDK. If you'd like to have a commit with your name on it, submit a PR to this branch in my repo (https://github.com/jnm2/nunit/tree/ui_message_pump_lightup) and I'll fast-forward merge to bring it into this PR. Or if you want to just paste an example, that's great too.

**Comment by batzen** [comment]
@jnm2 Created a PR in your fork.

**Comment by jnm2** [comment]
@batzen Thanks so much for solving that problem for us, and @mikkelbu thanks for reviewing!

**Comment by batzen** [comment]
@jnm2 you are welcome. thanks for solving the issue. Especially in such a future proof way.

**Comment by jnm2** [comment]
I forgot, but CI didn't. The same changes are needed for nunitlite-runner.csproj since it is the host process for the CI tests. On it.

**Pull request merged**